lib/jsonapi_compliable/errors.rb

```diff
@@ -1,7 +1,61 @@
 module JsonapiCompliable
   module Errors
     class Base < StandardError;end
-    class BadFilter < Base; end
+
+    class AttributeError < Base
+      attr_reader :resource,
+        :name,
+        :flag,
+        :exists,
+        :request,
+        :guard
+
+      def initialize(resource, name, flag, **opts)
+        @resource = resource
+        @name = name
+        @flag = flag
+        @exists = opts[:exists] || false
+        @request = opts[:request] || false
+        @guard = opts[:guard]
+      end
+
+      def action
+        if @request
+          {
+            sortable: 'sort on',
+            filterable: 'filter on'
+          }[@flag]
+        else
+          {
+            sortable: 'add sort',
+            filterable: 'add filter'
+          }[@flag]
+        end
+      end
+
+      def resource_name
+        name = if @resource.is_a?(JsonapiCompliable::Resource)
+          @resource.class.name
+        else
+          @resource.name
+        end
+        name || 'AnonymousResourceClass'
+      end
+
+      def message
+        msg = "#{resource_name}: Tried to #{action} on attribute #{@name.inspect}"
+        if @exists
+          if @guard
+            msg << ", but the guard #{@guard.inspect} did not pass."
```

Comment:
Where are we expecting these error messages to surface?  At some point I think errors that cause requests too return an error status will need to surface a helpful error message to the user, but these messages provide too many specifics around implementation details of a specific API.  If this is only meant to be shown to developers of the API, then 👍, but we should make sure we don't accidentally expose that down the line with our "friendly errors" work toward the end of this effort.

Probably turns into something like `def user_message` versus `def message` in the end, with showing the real message being controlled by `show_raw_errors`.

Reply:
We'd use the same error handling pattern, which would avoid showing these messages to the user (you'd get a 500). Users can rescue the error and customize the response, and/or we can register some defaults. Either way, I think it's orthagonal to these errors which are meant to be descriptive as possible for debugging.